pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,344 @@
+ * | MaximumPinotConsumptionDelayTracker|           ...
+ * |____________________________________|<-recordPinotConsumptionDelay()-{LLRealtimeSegmentDataManager (Partition n}}
+ *              ^                      ^
+ *              |                       \
+ *   tickInactivePartitions()    stopTrackingPinotConsumptionDelay(partitionId)
+ *    _________|__________                \
+ *   | TimerTrackingTask |          (CONSUMING -> DROPPED state change)
+ *   |___________________|          (CONSUMING -> OFFLINE state change)
+ *
+ */
+
+public class MaximumPinotConsumptionDelayTracker {
+  /*
+   * Class to handle timer thread that will track inactive partitions
+   */
+  private class TrackingTimerTask extends TimerTask {
+    private String _name;
+    private MaximumPinotConsumptionDelayTracker _tracker;
+
+    public TrackingTimerTask(String name, MaximumPinotConsumptionDelayTracker tracker) {
+      _name = name;
+      _tracker = tracker;
+    }
+
+    @Override
+    public void run() {
+      // tick inactive partitions every interval of time to keep tracked partitions up to date.
+      _tracker.tickInactivePartitions();
+    }
+  }
+
+  /*
+   * Class to keep a Pinot Consumption Delay measure and the time when the sample was taken (i.e. sample time)
+   * We will use the sample time to increase consumption delay when a partition stops consuming: the time
+   * difference between the sample time and current time will be added to the metric when read.
+   */
+  private class DelayMeasure {
+    public DelayMeasure(long t, long d) {
+      _delayMilliseconds = d;
+      _sampleTime = t;
+    }
+    public long _delayMilliseconds;
+    public long _sampleTime;
+  }
+
+  // HashMap used to store delay measures for all partitions active for the current table.
+  private HashMap<Long, DelayMeasure> _partitionToDelaySampleMap = new HashMap<>();
+  // We mark partitions that go from CONSUMING to ONLINE in _partitionsMarkedForConfirmation: if they do not
+  // go back to CONSUMING in some timeout, we confirm they are still hosted in this server by reading
+  // ideal state. This is done with the goal of minimizing reading ideal state for efficiency reasons.
+  // We do tick up counts associated with inactive partitions so that we can verify they are still
+  // hosted by the current server after a preconfigured number of ticks.
+  private HashMap<Long, Integer> _partitionsMarkedForConfirmation = new HashMap<>();
+  // Check ideal state only after _MAX_NUMBER_OF_TIME_TICKS_BEFORE_VERIFICATION * _TIMER_THREAD_TICK_INTERVAL_MS
+  private static final int MAX_NUMBER_OF_TIME_TICKS_BEFORE_VERIFICATION = 5;
+  // Sleep interval for timer thread that triggers read of ideal state
+  private static final int TIMER_THREAD_TICK_INTERVAL_MS = 60000;
+  // Delay Timer thread for this time after starting timer
+  private static final int INITIAL_TIMER_THREAD_DELAY_MS = 1000;
```

Comment:
can you add some comments as to why these values for timers are what they are?

Reply:
Done